src/librustc_trans/partitioning.rs

```diff
@@ -0,0 +1,376 @@
+            name: codegen_unit.name.clone(),
+            items: FnvHashMap(),
+        };
+
+        // Add all translation items that are not already there
+        for trans_item in reachable {
+            if let Some(linkage) = codegen_unit.items.get(&trans_item) {
+                // This is a root, just copy it over
+                final_codegen_unit.items.insert(trans_item, *linkage);
+            } else {
+                if initial_partitioning.roots.contains(&trans_item) {
+                    // This item will instantiated in some other codegen unit,
+                    // so we just add it here with AvailableExternallyLinkage
+                    final_codegen_unit.items.insert(trans_item, llvm::AvailableExternallyLinkage);
+                } else {
+                    // We can't be sure if this will also be instantiated
+                    // somewhere else, so we add an instance here with
+                    // LinkOnceODRLinkage. That way the item can be discarded if
+                    // it's not needed (inlined) after all.
+                    final_codegen_unit.items.insert(trans_item, llvm::LinkOnceODRLinkage);
+                }
+            }
+        }
+
+        final_partitioning.push(final_codegen_unit);
+    }
+
+    return final_partitioning;
+
+    fn follow_inlining<'tcx>(trans_item: TransItem<'tcx>,
+                             inlining_map: &InliningMap<'tcx>,
+                             visited: &mut FnvHashSet<TransItem<'tcx>>) {
+        if !visited.insert(trans_item) {
+            return;
+        }
+
+        if let Some(inlined_items) = inlining_map.get(&trans_item) {
+            for &inlined_item in inlined_items {
+                follow_inlining(inlined_item, inlining_map, visited);
+            }
+        }
+    }
+}
+
+fn characteristic_def_id_of_trans_item<'tcx>(tcx: &TyCtxt<'tcx>,
+                                             trans_item: TransItem<'tcx>)
+                                             -> Option<DefId> {
+    match trans_item {
+        TransItem::Fn(instance) => {
+            // For now, just use the actual DefId of the item, with the
+            // consequence that, for example, methods are emitted in the
+            // module where the impl is located instead of the one where
+            // the self-type is located.
+            Some(instance.def)
+        }
+        TransItem::DropGlue(t) => characteristic_def_id_of_type(t),
+        TransItem::Static(node_id) => Some(tcx.map.local_def_id(node_id)),
+    }
+}
+
+fn compute_codegen_unit_name<'tcx>(tcx: &TyCtxt<'tcx>,
+                                   def_id: DefId,
+                                   volatile: bool)
+                                   -> InternedString {
+    // Unfortunately, TyCtxt::with_path() is a bit too clever for us here,
+    // because will make a difference between directly and indirectly
+    // referenced crates. We don't want that.
+    let mut mod_path = String::with_capacity(64);
```

Comment:
The code in `librustc/ty/item_path` was, fwiw, intentionally written to be customizable for cases like this, by supplying a distinct `ItemPathBuffer`. Perhaps we should write a helper that finds the def-id of containing module, and then just use (some variant on) the `item_path_str` code?

idk, this fn is fairly simple, but I wanted to avoid a proliferation of misc bits of code that "push names onto paths"